docs/modules/ROOT/pages/compact-binary-specification.adoc

```diff
@@ -0,0 +1,706 @@
+
+Each type supported in the wire format has its type ID. The type IDs are used while constructing the schemas, performing type checks when accessing to fields, and are exposed as a public API.
+
+=== Nullable Primitives
+
+They are implemented as variable-sized types. The `null` values of such types are represented exactly as `null` variable-sized fields, with the offset of `-1` and no data.
+
+== Header
+
+The partition hash and the type ID are common for all serialization methods supported by Hazelcast. Therefore, the new format is no exception and every serialized object has a header in addition to the payload on the wire.
+
+[cols="1,1,3"]
+|===
+|Name |Type |Description
+|Partition hash |i32 |`BIG_ENDIAN` integer, used for key objects. Not applicable to value objects.
+|Type ID |i32 |`BIG_ENDIAN` integer that determines the serializer to be used. -55 for compact.
+|===
+
+== Var-Size Objects
+
+In this section, how a user-defined type is represented at the wire level is described. The binary consists of `Header`, `Data`, and `Offsets` sections in this order.
+
+=== Header Section
+
+[cols="1,1,1"]
+|===
+|Name |Type |Description
+|Schema ID |i64 |	
+Hash of the schema.
+|Data length |i32 |Length of the DATA SECTION below.
+|===
+
+=== Data Section
+
+[cols="1,1"]
+|===
+|Name |Description
+|Fixed-size Fields | Offsets of these fields will be deduced from the schema.
+|Variable-size Fields | 
+|===
+
+=== Offsets Section
+
+[cols="1,1,3"]
+|===
+|Name |Type |Description
+|Variable-Size FieldOffset index 0 |u8/u16/i32 |The index of a field offset is written in the Schema. Offsets of variable length fields. -1 for null
+|Variable-Size FieldOffset index 1 |u8/u16/i32 |The index of a field offset is written in the Schema. Offsets of variable length fields. -1 for null
+|... |... | 
+|Variable-Size FieldOffset index n |u8/u16/i32 |The index of a field offset is written in the Schema. Offsets of variable length fields. -1 for null
+|===
+
+Note that if the composed data does not include any variable-size field in the schema, `Variable-Size FieldOffset` and `DataLength` will not exist on the wire.
+
+Similarly, if there is no fixed-size field in the schema, `Fixed-Size Fields` will not exist on the wire.
+
+`Variable-Size FieldOffsets` are calculated from the beginning of the `DATA SECTION` shown in the table above.
+
+`Variable-Size FieldOffset` sizes vary depending on the Data Length.
+
+* Data Length <= `254`, offsets are `u8` (`255` is reserved for `null`)
+* Data Length <= `65534`, offsets are `u16` (`65535` is reserved for `null`)
+* Otherwise, offsets are `i32`.
+
+Length is written before offsets so that the binary can be skipped even when the schema cannot be found.
+
+A Variable-Size FieldOffset is `-1` if a Variable-Size field is `null`.
+
+Fixed-Size Fields cannot be `null`.
+
+== Fixed-Size Fields
+
+The fixed-size fields are written right after the `Length` field consecutively. They are accessed via `offset` written in the Schema.
```

Comment:
Partially applied, the offset is not given in the schema it is calculated from. Made second sentence 'Fixed-size fields are accessed via their offsets calculated from the schema'

Reply:
@srknzl Agreed. Please also add "that are" between "offsets" and "calculated", and a full stop at the end of the sentence. Thank you.

Reply:
there was a full stop, I added "that are" (btw if you see my comments right after I write them, I may not have pushed the changes yet, it will be eventually there though)